tests/integration/synapseclient/core/test_download.py

```diff
@@ -475,6 +475,80 @@ def mock_httpx_send(**kwargs):
         assert file.file_handle.content_md5 == file_md5
         assert os.path.exists(file_path)
 
+    async def test_download_from_url_multi_threaded_random_protocol_exceptions(
+        self,
+        syn: Synapse,
+        project_model: Project,
+        schedule_for_cleanup: Callable[..., None],
+    ) -> None:
+        """Test download of a file if downloaded in multiple parts. In this case I am
+        dropping the download part size to 500 bytes to force multiple parts download.
+
+        This function will randomly fail the download of a part with protocol exceptions
+
+        """
+        # Set the failure rate to 90%
+        failure_rate = 0.9
+
+        # GIVEN a file stored in synapse
+        file_path = utils.make_bogus_data_file()
+        file = await File(path=file_path, parent_id=project_model.id).store_async()
+        schedule_for_cleanup(file.id)
+        schedule_for_cleanup(file_path)
+        file_md5 = file.file_handle.content_md5
+        assert file_md5 is not None
+        assert os.path.exists(file_path)
+
+        # AND the file is not in the cache
+        syn.cache.remove(file_handle_id=file.file_handle.id)
+        os.remove(file_path)
+        assert not os.path.exists(file_path)
+
+        # AND an httpx client that is not mocked
+        httpx_timeout = httpx.Timeout(70, pool=None)
+        client = httpx.Client(timeout=httpx_timeout)
+
+        # AND the mock httpx send function to simulate a failure
+        def mock_httpx_send(**kwargs):
+            """Conditionally mock the HTTPX send function to simulate a failure. The
+            HTTPX .stream function internally calls a non-contexted managed send
+            function. This allows us to simulate a failure in the send function."""
+            is_part_stream = False
+            for header in kwargs.get("request").headers.raw:
+                if header[0].lower() == b"range":
+                    is_part_stream = True
+                    break
+            if is_part_stream and random.random() <= failure_rate:
```

Comment:
Why is it okay to have a random component to this test?

Reply:
This was to prove that in adverse conditions we don't have direct control over that the code still behaves as expected. The behavior is that it would retry up to 5 minutes to complete the transfer, after which it would fail.

Since the exception is being raised without making any HTTP call the fail -> retry loop is very quick. In this test I set a maximum backoff of 0.2 seconds between retries.


I used this same kind of testing to verify what would occur before the download re-write changes, to post re-write. I found the new code to work 100% of the time, whereas the previous code worked ~20% of the time.

Reply:
With the failure rate what it is, you're practically guaranteed a failure. But wouldn't it be better to make the test entirely deterministic? Or would that require jumping through too many hoops?

Reply:
> With the failure rate what it is, you're practically guaranteed a failure

Yes, that is the purpose of having the failure rate what it is. It is meant to verify that the retry logic is sufficient to handle 90% of the parts of a file being dropped.

> But wouldn't it be better to make the test entirely deterministic? Or would that require jumping through too many hoops?

Integration tests are rarely 100% deterministic since it relies on so many outside factors. This is to help simulate if one of those outside factors (The remote file server) is one of those flaky items.

Since this gives the retry logic more testing I am keen to keep this as is.